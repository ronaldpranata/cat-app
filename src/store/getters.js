export default {
  getCatBreedColumn() {
    return ['Image', 'Name'];
  },
  getCatBreed(state) {
    return state.catBreeds;
  },
  getCatDetail(state) {
    return state.catDetail;
  },
  getOriginMap(state) {
    return state.catOriginMap;
  },
};
