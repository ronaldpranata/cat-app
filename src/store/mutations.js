export default {
  SET_CAT_BREEDS: (state, data) => {
    state.catBreeds = data;
  },
  SET_MAP_LOCATION: (state, data) => {
    state.catOriginMap = data;
  },
};
