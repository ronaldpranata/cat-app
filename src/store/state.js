export default () => ({
  catDetail: null,
  catBreeds: [],
  catOriginMap: null,
});
