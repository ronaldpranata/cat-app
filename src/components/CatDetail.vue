<template>
  <div class="cat">
    <a class="bold clickable" @click="$router.go(-1)">&laquo; back</a><br />
    <h1 class="title">{{ t('catDetail') + " " + cat?.name }}</h1>
    <img :src="cat?.image?.url" class="cat-image" />
    <table>
      <tr>
        <td class="bold">{{ t('description') }}</td>
        <td class="detail description" v-html="cat?.description"></td>
      </tr>
      <tr>
        <td class="bold">{{ t('maximumWeight') }}</td>
        <td class="detail weight">
          <div v-for="(weight, key) in cat?.weight" :key="key">
            {{ key }} : {{ weight }}
          </div>
        </td>
      </tr>
      <tr>
        <td class="bold">{{ t('averageLifespan') }}</td>
        <td class="detail life-span">{{ cat?.life_span }}</td>
      </tr>

      <tr>
        <td class="bold">{{ t('origin') }}</td>
        <td class="detail gmap">
          <GMapMap :center="center" v-if="center" :zoom="3" map-type-id="terrain" style="width: 500px; height: 300px">
            <GMapCluster v-if="markers.length>0">
              <GMapMarker :key="index" v-for="(m, index) in markers" :position="m.position" :clickable="true"
                :draggable="true" @click="center = m.position" />
            </GMapCluster>
          </GMapMap>
        </td>
      </tr>
    </table>
  </div>
</template>

<script>
import { defineComponent } from 'vue';
import { useI18n } from 'vue-i18n';
import { GET_ORIGIN_LOCATION } from '@/store/action_types';

export default defineComponent({
  name: 'CatDetail',
  computed: {
    center() {
      return this.$store.getters.getOriginMap?.results[0]?.geometry?.location;
    },
    markers() {
      return [
        {
          position: this.center,
        },
      ];
    },
  },
  props: {
    cat: Object,
  },
  setup() {
    const { t } = useI18n();
    return { t };
  },
  async created() {
    await this.$store.dispatch(GET_ORIGIN_LOCATION, { address: this.cat.origin });
  },
});
</script>
<style scoped lang='scss'>
.cat {
  max-width: 700px;
  margin: auto;

  .cat-image {
    max-width: 200px;
  }

  .bold {
    font-weight: bold;
  }

  .detail {
    text-align: left;
  }
}
</style>
