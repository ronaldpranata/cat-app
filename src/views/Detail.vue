<template>
  <div class="detail">
    <CatDetail :cat="catDetail" />
  </div>
</template>

<script>
import { useI18n } from 'vue-i18n';
import CatDetail from '@/components/CatDetail.vue';

export default {
  name: 'Detail',
  components: {
    CatDetail,
  },
  setup() {
    const { t } = useI18n();
    return { t };
  },
  computed: {
    catDetail() {
      return this.$store.getters.getCatDetail;
    },
  },
  async mounted() {
    if (!this.catDetail) {
      this.$router.push({ path: '/' });
    }
  },
};
</script>
