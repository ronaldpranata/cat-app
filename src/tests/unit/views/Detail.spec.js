import { mount } from '@vue/test-utils';
import Vuex from 'vuex';
import Detail from '@/views/Detail.vue';
import CatDetail from '@/components/CatDetail.vue';
import mockCatBreeds from '@/__mock__/mock';
import { createI18n } from 'vue-i18n';
import state from '@/store/state';
import mutations from '@/store/mutations';
import getters from '@/store/getters';
import {
  GET_CAT_BREEDS,
} from '@/store/action_types';

const i18n = createI18n({
  // vue-i18n options here ...
});

const actions = {
  async [GET_CAT_BREEDS]({ commit }) {
    try {
      const response = await new Promise((resolve) => {
        resolve(mockCatBreeds);
      });
      commit('SET_CAT_BREEDS', response?.data);
      return response?.data;
    } catch (error) {
      throw new Error(error);
    }
  },
};
const store = new Vuex.Store({
  state,
  actions,
  mutations,
  getters,
});

const factory = () => mount(Detail, {
  global: {
    plugins: [i18n, Vuex],
    mocks: {
      $store: store,
    },
  },
});

describe('Detail', () => {
  test('renders a CatDetail', () => {
    const wrapper = factory();
    const component = wrapper.findComponent(CatDetail);
    expect(component.exists()).toBeTruthy();
  });

  test('mounts properly', () => {
    const wrapper = factory();
    expect(wrapper.vm).toBeTruthy();
  });

  test('renders properly', () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
