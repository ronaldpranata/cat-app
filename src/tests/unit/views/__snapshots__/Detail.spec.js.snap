// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail renders properly 1`] = `
<div class="detail">
  <div class="cat"><a class="bold clickable">Â« back</a><br>
    <h1>catDetail undefined</h1><img class="cat-image">
    <table>
      <tr>
        <td class="bold">description</td>
        <td class="detail"></td>
      </tr>
      <tr>
        <td class="bold">maximumWeight</td>
        <td class="detail"></td>
      </tr>
      <tr>
        <td class="bold">averageLifespan</td>
        <td class="detail"></td>
      </tr>
      <tr>
        <td class="bold">origin</td>
        <td class="detail">
          <!--v-if-->
        </td>
      </tr>
    </table>
  </div>
</div>
`;
